File: generator/generate_hydrostatic.cpp
#include "../src/particle.hpp"
#include "../src/particles.hpp"
#include "../src/particles_exporter.hpp"
#include "generator_src/generator_dialogue.hpp"

#include <algorithm>
#include <array>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>
namespace fs = std::filesystem;

class Box2d {
private:
    Eigen::Vector2d center;
    Eigen::Vector2d box_size;

public:
    Box2d() = default;
    Box2d(double xMin, double xMax, double yMin, double yMax) {
        center << (xMin + xMax) / 2.0, (yMin + yMax) / 2.0;
        box_size << xMax - xMin, yMax - yMin;
    }
    double sdf(const Eigen::Vector2d& r) const {
        auto q = (r - center).array().abs() - box_size.array() / 2.0;
        return q.max(0.0).matrix().norm() + std::min(std::max(q.x(), q.y()), 0.0);
    }
};

int main(int argc, char** argv) {
    Particles particles;

    double l0      = 0.012;
    double density = 1000.0;

    Box2d fluidDomain(-0.18, 0.18, 0, 0.48);
    Box2d emptyDomain(-0.18, 0.18, 0.48, 0.6);
    Box2d wallBB(-0.18 - 2 * l0, 0.18 + 2 * l0, -2 * l0, 0.6);
    Box2d dummyBB(-0.18 - 4 * l0, 0.18 + 4 * l0, -4 * l0, 0.6);

    int id = 0;
    for (int ix = (int) round(-0.18 / l0) - 4; ix < (int) round(0.18 / l0) + 4; ix++) {
        for (int iy = (int) round(0 / l0) - 4; iy < (int) round(0.6 / l0); iy++) {
            Eigen::Vector2d r(l0 * (ix + 0.5), l0 * (iy + 0.5));
            Eigen::Vector3d r_3d(r.x(), r.y(), 0);

            // empty region
            if (emptyDomain.sdf(r) < 0) {
                continue;
            }
            // fluid region
            if (fluidDomain.sdf(r) < 0) {
                particles.add(Particle(id, ParticleType::Fluid, r_3d, Eigen::Vector3d::Zero(), density));
                id++;
                continue;
            }
            // wall region
            if (wallBB.sdf(r) < 0) {
                particles.add(Particle(id, ParticleType::Wall, r_3d, Eigen::Vector3d::Zero(), density));
                id++;
                continue;
            }
            // dummy region
            if (dummyBB.sdf(r) < 0) {
                particles.add(Particle(id, ParticleType::DummyWall, r_3d, Eigen::Vector3d::Zero(), density));
                id++;
                continue;
            }
        }
    }

    ParticlesExporter exporter;
    exporter.setParticles(particles);
    std::string parentPath = "input/hydrostatic";
    GeneratorDialogue gd;
    gd.generatorDialogue(
        fs::path(parentPath), 
        particles, 
        {".prof", ".vtu"}
    );  
}

File: generator/generate_hydrostatic.py
import numpy as np
import meshio
from pathlib import Path


class Box2d:
    def __init__(self, x_min, x_max, y_min, y_max):
        self.senter = np.array([(x_min + x_max) / 2, (y_min + y_max) / 2])
        self.size = np.array([x_max - x_min, y_max - y_min])

    def sdf(self, position):
        q = np.abs(position - self.senter) - self.size / 2
        return np.linalg.norm(np.maximum(q, 0)) + np.min(np.maximum(q, 0))


def generate(l0, prof_path, vtu_path):
    fluid_domain = Box2d(-0.18, 0.18, 0, 0.48)
    empty_domain = Box2d(-0.18, 0.18, -0.48, 0.6)
    wall_domain = Box2d(-0.18 - 2 * l0, 0.18 + 2 * l0, -2 * l0, 0.6)
    dummy_domain = Box2d(-0.18 - 4 * l0, 0.18 + 4 * l0, -4 * l0, 0.6)

    positions = []
    types = []

    for ix in range(round(-0.18 - 4 * l0), round(0.18 / l0) + 4):
        for iy in range(round(-4 * l0), round(0.6 / l0) + 4):
            pos_i = np.array([ix * l0, iy * l0])

            if empty_domain.sdf(pos_i) < 0:
                continue
            elif fluid_domain.sdf(pos_i) < 0:
                positions.append(pos_i)
                types.append(0)
            elif wall_domain.sdf(pos_i) < 0:
                positions.append(pos_i)
                types.append(1)
            elif dummy_domain.sdf(pos_i) < 0:
                positions.append(pos_i)
                types.append(2)

    # write prof file
    # write vtu file with meshio


if __name__ == "__main__":
    l0 = 0.012
    project_dir = Path(__file__).parent.parent
    dst_dit = project_dir / "input" / "hydrostatic"
